File: lib/to_mdx/parsers/open_document/draw/custom_shape.rb
# frozen_string_literal: true

require_relative "../base.rb"
require_relative "../../../detectors/code_detector"
require_relative "../mixins/positionable_element"
require_relative "../mixins/svg"

module ToMdx
  module Parsers
    module Draw
      class CustomShape < Base
        include ToMdx::Detectors::CodeDetector
        include Mixins::PositionableElement
        include Mixins::Svg

        def initialize(element = nil, context = {})
          super(element, context)
          @draw_style_name = context[:styles].use(element&.attributes&.[]("draw:style-name"))
        end

        def to_mdx
          class_attr = @draw_style_name.empty? ? "" : " className=\"#{@draw_style_name}\""

          code_snippet = maybe_code_snippet_mdx
          return "<div#{class_attr} style={#{generate_combined_style_object}}>#{code_snippet}</div>" if code_snippet

          <<-MDX
            <div#{class_attr} style={#{generate_combined_style_object}}>
              #{contentful_children.map(&:to_mdx).compact.join}
            </div>
          MDX
        end

        def to_s
          contentful_children.map(&:to_s).compact.join
        end

        def with_context(context)
          context.merge(
            parent_style_name: @draw_style_name,
            parent_width: @width,
            parent_height: @height,
            parent_is_code_snippet: is_using_monospace_font?
          )
        end

        register_for "draw:custom-shape"
      end
    end
  end
end

File: src/parsers/open-document/draw/custom_shape.ts
import dedent from "dedent-js";
import json5 from "json5";
import { CodeDetector } from "../../../detectors/code-detector.ts";
import type { Style } from "../../../styles.ts";
import type { Maybe } from "../../../utils.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";
import { FormulaEvaluator } from "../mixins/formula_evaluator.ts";
import { LayoutDetection } from "../mixins/layout_detection.ts";
import { SvgElement } from "../mixins/svg_element.ts";
import { EnhancedGeometry } from "./enhanced_geometry.ts";

declare module "../../base_element.ts" {
  interface ParseContext {
    isSvg?: boolean;
    formulaEvaluator?: FormulaEvaluator;
  }
}

export class CustomShape extends SvgElement {
  readonly formulaEvaluator: FormulaEvaluator;

  private drawStyle: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.drawStyle = context.styles.use(this.attr("draw:style-name"));
    this.formulaEvaluator = new FormulaEvaluator(this);
  }

  toMdx(): string {
    const layoutDetection = new LayoutDetection(this);
    const codeSnippet = this.codeDetector.maybeCodeSnippet();
    if (codeSnippet) {
      if (layoutDetection.needsPositioning) {
        return dedent`
          <div style={${json5.stringify(layoutDetection.generatePositioningStyleObject())}}>
            ${codeSnippet}
          </div>
        `;
      }
      return codeSnippet;
    }

    let content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join("");

    let divClassAttr = "";
    if (this.isSvg) {
      content = this.wrapWithSvgTag(content, this.drawStyle);
    } else {
      divClassAttr = ` className="${this.drawStyle}"`;
    }

    if (!layoutDetection.needsPositioning) {
      return content;
    }

    const styleObject = layoutDetection.generatePositioningStyleObject();
    return dedent`
      <div${divClassAttr} style={${json5.stringify(styleObject, { quote: '"' })}}>
        ${content}
      </div>
    `;
  }

  toString(): string {
    return this.contentfulChildren()
      .map((child) => child.toString())
      .filter((content) => content.trim().length > 0)
      .join("");
  }

  get layoutClass(): Maybe<string> {
    return new LayoutDetection(this).layoutClass;
  }

  get isSvg(): boolean {
    const enhancedGeometry = this.children.filter((child) => child instanceof EnhancedGeometry);
    switch (enhancedGeometry.length) {
      case 0:
        return false;
      case 1:
        // In this case, we'll assume this is just a border
        return (enhancedGeometry[0] as unknown as EnhancedGeometry).drawType !== "ooxml-rect";
      default:
        return true;
    }
  }

  protected withContext(context: ParseContext): ParseContext {
    const newContext = {
      ...context,
      formulaEvaluator: new FormulaEvaluator(this),
      parentWidth: parseFloat(this.width || "0"),
      parentHeight: parseFloat(this.height || "0"),
    };

    Object.defineProperty(newContext, "viewBox", { get: () => this.viewBox });
    Object.defineProperty(newContext, "isSvg", { get: () => this.isSvg });

    return newContext;
  }

  private get codeDetector(): CodeDetector {
    return new CodeDetector(this, () => this.toString());
  }

  static {
    BaseElement.registerFor.call(this, "draw:custom-shape");
  }
}
